src/hooks/useNetworkGraph.ts

```diff
@@ -0,0 +1,206 @@
+  mainNode: string;
+  limitNestingLevel: boolean;
+  nestingLevel: number;
+  searchValueLabel?: string;
+  searchValueSubLabel?: string;
+  onEdgeSelected?: (node1: any, node2: any) => void;
+}
+
+export const useNetworkGraph = ({
+  sigmaContainerId = 'sigma-container',
+  mainNode,
+  limitNestingLevel,
+  nestingLevel,
+  searchValueLabel,
+  searchValueSubLabel,
+  onEdgeSelected,
+}: IUseNetworkGraphProps) => {
+  const graph = useRef<Graph>();
+  const renderer = useRef<Sigma>();
+
+  const [isGraphRendered, setIsGraphRendered] = useState(false);
+
+  const [hoveredNode, setHoveredNode] = useState<string>();
+  const [selectedEdge, setSelectedEdge] = useState<string>();
+  const selectedEdgeRef = useRef<string>();
+
+  const navigate = useNavigate();
+
+  const selectEdge = (milestone1?: string, milestone2?: string) => {
+    if (milestone1 && milestone2) {
+      setSelectedEdge(graph?.current?.edge(milestone1, milestone2));
+    } else {
+      setSelectedEdge(undefined);
+    }
+  };
+
+  const createNetworkGraph = useCallback(
+    (onGraphCreated: OnGraphCreatedFunction) => {
+      const sigmaContainer = document.getElementById(sigmaContainerId);
+
+      if (sigmaContainer && !graph.current) {
+        graph.current = new Graph();
+
+        renderer.current = new Sigma(graph.current, sigmaContainer, {
+          minCameraRatio: 0.1,
+          maxCameraRatio: 10,
+          renderEdgeLabels: true,
+          enableEdgeHoverEvents: true,
+          enableEdgeClickEvents: true,
+          labelRenderer: drawLabel,
+        });
+
+        renderer.current.setSetting(
+          'hoverRenderer',
+          (context: CanvasRenderingContext2D, data: PlainObject, settings: PlainObject) =>
+            drawHover(context, { ...renderer.current?.getNodeDisplayData(data.key), ...data }, settings)
+        );
+
+        // EVENT HANDLERS
+
+        renderer.current.on('enterNode', ({ node }) => {
+          setHoveredNode(node);
+          sigmaContainer.style.cursor = 'pointer';
+        });
+
+        renderer.current.on('leaveNode', () => {
+          setHoveredNode(undefined);
+          sigmaContainer.style.cursor = 'default';
+        });
+
+        renderer.current.on('clickNode', ({ node }) => {
+          navigate(graph.current?.getNodeAttributes(node)?.link);
+        });
+
+        renderer.current?.on('enterEdge', () => {
+          sigmaContainer.style.cursor = 'pointer';
+        });
+
+        renderer.current?.on('leaveEdge', () => {
+          sigmaContainer.style.cursor = 'default';
+        });
+
+        // ON GRAPH CREATED
+
+        onGraphCreated(graph.current, renderer.current);
+
+        circular.assign(graph.current);
+        const settings = forceAtlas2.inferSettings(graph.current);
+        forceAtlas2.assign(graph.current, { settings, iterations: 500 });
+
+        setIsGraphRendered(true);
+      }
+    },
+    [sigmaContainerId, navigate]
+  );
+
+  // DYNAMIC EVENT HANDLERS
+
+  useEffect(() => {
```

Comment:
Do agree that we need to use as less `useEffect` as possible unless there is no better substitution. we can have a try to see if it can be done within `onGraphCreated` or using `createNetworkGraph`.

Reply:
This is correct way to do it. And moving it elsewhere would solve nothing.